src/trace/trace.cpp

```diff
@@ -677,30 +683,55 @@ void Trace::merge_ips(IpRefMap& new_children, IpCctxMap& children,
         const auto& cctx = cctx_it->second.cctx;
         mapping_table.at(local_ref) = cctx.ref();
 
-        merge_ips(local_children, cctx_it->second.children, mapping_table, cctx, infos);
+        merge_ips(local_children, cctx_it->second.children, mapping_table, cctx, infos, pid);
     }
 }
 
-otf2::definition::mapping_table Trace::merge_calling_contexts(
-    IpRefMap& new_ips,
-    const std::unordered_map<pid_t, otf2::definition::calling_context::reference_type>& thread_refs,
-    std::map<pid_t, ProcessInfo>& infos)
+otf2::definition::mapping_table Trace::merge_calling_contexts(ThreadCctxRefMap& new_ips,
+                                                              size_t num_ip_refs,
+                                                              std::map<pid_t, ProcessInfo>& infos)
 {
 #ifndef NDEBUG
-    std::vector<uint32_t> mappings(new_ips.size() + thread_refs.size(), -1u);
+    std::vector<uint32_t> mappings(num_ip_refs, -1u);
 #else
-    std::vector<uint32_t> mappings(new_ips.size() + thread_refs.size());
+    std::vector<uint32_t> mappings(num_ip_refs);
 #endif
 
-    if (!new_ips.empty())
+    // Merge local thread tree into global thread tree
+    for (auto& local_thread_cctx : new_ips)
     {
-        merge_ips(new_ips, calling_context_tree_, mappings, otf2::definition::calling_context(),
-                  infos);
-    }
-    if (!thread_refs.empty())
-    {
-        merge_thread_calling_contexts(thread_refs, mappings);
+        auto tid = local_thread_cctx.first;
+        auto local_ref = local_thread_cctx.second.entry.ref;
+
+        auto global_thread_cctx = calling_context_tree_.find(tid);
+
+        if (global_thread_cctx == calling_context_tree_.end())
+        {
+            if (tid != 0)
+            {
+                // Threads rarely have their own name, in that case use the processes name
+                auto process_cctx = calling_context_tree_.find(local_thread_cctx.second.pid);
+                if(process_cctx == calling_context_tree_.end())
+                {
+                        add_thread(tid, "<unknown>");
```

Comment:
```suggestion
                        add_thread(tid, "<unknown process>");
```

Or unknown thread? 😕

Reply:
unknown thread sounds better to me as we are dealing with the tid here